### PR TITLE
Fix 13777 - refine check for caching companion in sum mirror

### LINKED ISSUE #13777: ClassCastException when summoning Mirror for hierarchical sum compiled by 3.0.x
## Compiler version

3.1.0

## Minimized code

compile with Scala 3.0.2 to `out` dir
```scala
sealed trait Top
object Top // companion is necessary

case class Middle() extends Top with Bottom
sealed trait Bottom extends Top
```

compile with Scala 3.1.0 and with `out` on classpath
```scala
val m = summon[deriving.Mirror.SumOf[Top]]
```

## Output

runtime error:
```scala
java.lang.ClassCastException: Top$ cannot be cast to scala.deriving.Mirror$Sum
  ... 28 elided
```
This happens because `Top` is an eligible sum, and it has a companion object (which if compiled by Scala 3.1.0 would extend `deriving.Mirror.Sum`), so the companion object is used as the mirror and is cast to `Sum`.

## Expectation
prevent summoning mirrors for hierarchical sums with companion objects compiled before 3.1.0

### PR DESCRIPTION
**Pull request by bishabosha**
~~This also exposes a check to see if a file is a "lateCompile" in the current run, as a late-compiled symbol's type info is derived from source - aka a companion that is late-compiled would not subtype `Mirror.Sum`.~~

Edit: the above is no longer a part of these changes

cache the mirror in the companion if:
- the class is not scala 2 defined
  & the class has a companion
  & the class was either:
    - defined in source
    - from tasty and its companion is a subtype of Mirror.Sum

fixes #13777

### CODE REVIEW

**Comment by bishabosha** [comment]
I have no idea what's happening with the fs2 test in the community build - there are some comparison tests that are producing unexpected results - does not seem related to this

**Comment by bishabosha** [comment]
@smarter I have implemented the simpler version of a from-source check

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala [hunk 1]
```diff
@@ -112,7 +112,14 @@ object SymUtils:
           self.isCoDefinedGiven(res.typeSymbol)
       self.isAllOf(Given | Method) && isCodefined(self.info)
 
-    def useCompanionAsMirror(using Context): Boolean = self.linkedClass.exists && !self.is(Scala2x)
+    def useCompanionAsSumMirror(using Context): Boolean =
+      self.linkedClass.exists
+      && !self.is(Scala2x)
+      && (
+        // self is from source, or companion is a subtype of Sum
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/SymUtils.scala:119]: It'd be good to explain why for things which are not from source we need the more precise check (i.e., the difference in code generation between scala 3.0 and 3.1).

**Review by smarter — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:119]: It's more about the other way round: ideally we can just check is subclass (which is only true in scala 3.1+).  But we call this check before that parent is added, for code in compilation, so have to handle that specially.

Feels like the two new ones would subsume `!self.is(Scala2x)`, right?  (minor)

**Review by dwijnand — APPROVED** [review]
Thanks, Jamie, for picking this up.  Good to see it didn't spin out from the fix we had theorycrafted.  Good to go, for me.  I would hope this could make the 3.1.1 patch release.

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala (hunk 1 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala]: I have clarified the comment to explain why we check if its from source

**Pull request merged**